src/wmtk/operations/polygon_mesh/DeleteBubble.hpp

```diff
@@ -0,0 +1,49 @@
+#pragma once
+#include <wmtk/operations/TupleOperation.hpp>
+#include "AtomicOperation.hpp"
+#include "PolygonMeshOperation.hpp"
+
+namespace wmtk::operations {
+namespace polygon_mesh {
+class DeleteBubble;
+}
+
+template <>
+struct OperationSettings<polygon_mesh::DeleteBubble>
+{
+};
+
+namespace polygon_mesh {
+
+/**
+ * @class Atomic operation for removing a bubble component consisting of a single edge with two
+ * distinct endpoints and a single face.
+ */
+class DeleteBubble : public AtomicOperation
+{
+public:
+    // constructor for default factory pattern construction
+    DeleteBubble(Mesh& m, const Tuple& e, const OperationSettings<DeleteBubble>& settings);
+
+    std::string name() const override;
+
+    using PolygonMeshOperation::hash_accessor;
+
+    /**
+     * @brief Check the precondition that the tuple specifies a bubble component edge.
+     *
+     * @return true if the tuple specifies a bubble edge
+     * @return false otherwise
+     */
+    bool precondition();
```

Comment:
We've had some discussion about this before. Denis wants to have the precondition as part of the interface (which I second), but perhaps it would be good to put my thoughts on them in text.

I think there's a somewhat important distinction between the two types of preconditions that exist for the atomic operations; there are (a) preconditions that ensure manifold boundary vertices and (b) preconditions that ensure valid topology.

(a) **preconditions that ensure manifold boundary vertices**: `Splice` is always manifold on closed meshes, and the precondition only checks whether vertices remain manifold with respect to hole faces after the operation. `MakeHole` and `FillHole` are similar. The underlying representation can actually support these non manifold vertices, so this seems like a reasonable place for an invariant or a predicate in the `PolygonMesh` class to check if all vertices are manifold as this is actually an invariant that we want to maintain at all times. Using an `after` invariant would actually simplify and unify the implementation a lot.

(b) **preconditions that ensure valid topology**: `DeleteBubble`, on the other hand, can actually catastrophically break the underlying topology adjacency structure if the precondition is not satisfied. It thus breaks our guarantee that the `PolygonMesh` always has valid topology (which we assumed when we moved the validity checks from the base class to the debugging wrapper). I would be somewhat in favor of taking the (very simple and constant time) precondition out of the assert and just returning without doing anything if it is not satisfied.

Compound operations also have (c) precondition that simply ensure the operations do what they should (e.g., that `SplitFace` actually splits a face). These don't really preserve any invariants as they still do something topologically valid if the precondition is not satisfied (e.g., `SplitFace` will merge in a new edge somewhere using two splice operations). It may still be good to have the precondition somewhere to make the assumptions explicit and testable, but these sort of preconditions I am mostly fine with making protected and only using for asserts (as long as we ensure atomic operations never break topology).

In general, I don't have any real issue making any of the preconditions into invariants, but Daniele was also somewhat insistent that the invariants should not be used for checking operation validity. The operations that take two halfedges may also be an issue. Only the manifold vertex after invariant seems straightforward and reasonable to me.